Update json docs (#5590)
OpenLiberty/docs: modules/ROOT/pages/json-p-b.adoc

}
----

> == Thrid-party JSON-P and JSON-B providers

Reviewer: ```suggestion
== Third-party JSON-P and JSON-B providers
```

> In Open Liberty, you can choose to use the reference implementations of JSON-P and JSON-B by using the feature:jsonp[display=Jakarta JSON Processing feature], for JSON-P, or the feature:jsonb[display=Jakarta JSON Binding feature], for JSON-B.

Reviewer: ```suggestion
In Open Liberty, you can use the default reference implementations of JSON-P and JSON-B by using the feature:jsonp[display=Jakarta JSON Processing feature], for JSON-P, or the feature:jsonb[display=Jakarta JSON Binding feature], for JSON-B.
```

> If you perfer to use a thrid-party implementation for JSON-P and JSON-B, then you will want to instead use the feature:jsonpContainer[display=Jakarta JSON Processing Container feature], for JSON-P, or the feature:jsonbContainer[display=Jakarta JSON Binding Container feature], for JSON-B. Then, configure a library and a BELL (Basic Extension using Liberty Libraries) that provides an implementation of the JSON-P or JSON-B API.

Reviewer: ```suggestion
If you prefer to use a third-party implementation for JSON-P or JSON-B,  use the feature:jsonpContainer[display=Jakarta JSON Processing Container feature], for JSON-P, or the feature:jsonbContainer[display=Jakarta JSON Binding Container feature], for JSON-B instead. Then, configure a library and a Basic Extension using Liberty Libraries (BELL ) that provides an implementation of the JSON-P or JSON-B API. You can specify a third-party implementation for either the JSON-P or JSON-B API by configuring a config:bell[] element in your `server.xml` file.
```

> For example, Apache Johnzon is another JSON-B compliant implementation. 

Reviewer: ```suggestion
For example, [Apache Johnzon](https://johnzon.apache.org/) is a third-party JSON-B implementation. 
```

> In this example, a server is configured to use Johnzon instead of the reference implementation Yasson:

Reviewer: ```suggestion
In this example, a server is configured to use Johnzon instead of the default  reference implementation, which is Yasson:
```

[source, xml]
----
<server>
  <featureManager>
    <feature>jsonbContainer-2.0</feature>
  </featureManager>

  <bell libraryRef="johnzon"/>

  <library id="johnzon">
    <fileset dir="${server.config.dir}/johnzon"includes="*.jar"/>
  </library>
</server>
----

> You can also mix and match reference implementations and third-party implementations so long as both implmentations are written to support the same Jakarta EE version. 

Reviewer: ```suggestion
You can also mix and match reference implementations and third-party implementations, as long as both implementations support the same Jakarta EE version. 
```

> For example, Apache Johnzon is another JSON-P compliant implementation. 

Reviewer: ```suggestion
For example, Apache Johnzon also provides a JSON-P implementation. 
```

> In this example, a server is configured to use Johnzon instead of Glassfish for JSON-P, but continues to use Yasson for JSON-B:

Reviewer: ```suggestion
In this example, a server is configured to use Johnzon instead of the default reference implementation for JSON-P, which is Glassfish. However, the server continues to use Yasson for JSON-B:
```

[source, xml]
----
<server>
  <featureManager>
    <feature>jsonb-2.0</feature>
    <feature>jsonpContainer-2.0</feature>
  </featureManager>

  <bell libraryRef="johnzon"/>

  <library id="johnzon">
    <fileset dir="${server.config.dir}/johnzon" includes="*.jar"/>
  </library>
</server>
----

=== Avoiding trouble

> There are other features, such as microProfile, that enable the jsonp and jsonb features.

Reviewer: ```suggestion
Other features, such as feature:microProfile[display=MicroProfile], that enable the JSON-P and JSON-B features by default.
```

> When you want to use one of these features, and still provide your own thrid-party implementation for the JSON-P or JSON-B API you can. 

Reviewer: However, you can still still provide your own third-party implementation for the JSON-P or JSON-B API, even when the reference implementations are enabled by default.

> The jsonbContainer and jsonpContainer features will supersed the jsonb and jsonp features respectively. 

Reviewer: ```suggestion
The` jsonbContainer` and `jsonpContainer` features supersede the` jsonb` and `jsonp` features. Enabling either feature allows you to provide a third-party implementation.
```

> If logging is enabled you will see the following information message to confirm this behavior: 

Reviewer: ```suggestion
If logging is enabled, the following information message confirms this behavior.
```

> When you want to use one of these features, and still provide your own thrid-party implementation for the JSON-P or JSON-B API you can. 

Reviewer: ```suggestion
However, you can still still provide your own third-party implementation for the JSON-P or JSON-B API, even when the reference implementations are enabled by default.
```

[source, txt]
----
CWWKJ0350I: The jsonb and jsonbContainer features are both enabled. The jsonbContainer feature supersedes the jsonb feature.
CWWKJ0351I: The jsonp and jsonpContainer features are both enabled. The jsonpContainer feature supersedes the jsonp feature.
----

> If you fail to provide a BELL configuration element that provides an implementation of the JSON API used by your application, then you may see the following error when your application attempts to use the JSON API. 

Reviewer: ```suggestion
However, you still must configure a `bell` element that specifies the JSON API implementation that your application uses. If not, the following error occurs when your application attempts to use the JSON API. 
```
